cmd/test.go

```diff
@@ -0,0 +1,171 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+	"fmt"
+	"github.com/HotelsDotCom/flyte/template"
+	"github.com/HotelsDotCom/flyte/execution"
+	jsont "github.com/HotelsDotCom/flyte/json"
+	"io/ioutil"
+	"encoding/json"
+	"github.com/ghodss/yaml"
+	"errors"
+	"path/filepath"
+)
+
+var format string
+
+func newTestCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "test file",
+		Short: "Test step execution with trigger event and optional context",
+		Long:  testCmdLong,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return errors.New("you need to provide a test file")
+			}
+			return nil
+		},
+		RunE: func(c *cobra.Command, args []string) error {
+			output, err := runTestCmd(args[0], format)
+			if err != nil {
+				return err
+			}
+			_, err = fmt.Fprint(c.OutOrStdout(), output)
+			return err
+		},
+	}
+
+	cmd.Flags().StringVarP(&format, "format", "f", "json", "Output format. One of: json|yaml")
+	return cmd
+}
+
+func runTestCmd(testFilePath, format string) (string, error) {
+	var step testStep
+	if err := unmarshalFile(testFilePath, &step); err != nil {
+		return "", err
+	}
+
+	action, err := step.execute()
+	if err != nil {
+		return "", err
+	}
+
+	out, err := marshal(action, format)
+	if err != nil {
+		return "", err
+	}
+
+	return string(out), nil
+}
+
+type testStep struct {
+	Step      execution.Step
+	Event     execution.Event
+	Context   map[string]string
+	Datastore map[string]interface{}
+}
+
+func (t testStep) execute() (*testAction, error) {
+	//override default datastore func which is using mongo to get data item
+	//use static map instead which can be passed in the input file
+	template.AddStaticContextEntry("datastore", datastoreFn(t.Datastore))
+
+	action, err := t.Step.Execute(t.Event, t.Context)
+	if err != nil {
+		return nil, err
+	}
+	return &testAction{
+		Name:       action.Name,
+		PackName:   action.PackName,
+		PackLabels: action.PackLabels,
+		Input:      action.Input,
+		Context:    action.Context,
+	}, nil
+}
+
+type testAction struct {
+	Name       string            `json:"name"`
+	PackName   string            `json:"packName"`
+	PackLabels map[string]string `json:"packLabels,omitempty"`
+	Input      jsont.Json        `json:"input,omitempty"`
+	Context    map[string]string `json:"context,omitempty"`
+}
+
+func unmarshalFile(filename string, v interface{}) error {
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return err
+	}
+
+	ext := filepath.Ext(filename)
+	switch ext {
+	case ".json":
+		return json.Unmarshal(data, v)
+	case ".yaml", ".yml":
+		return yaml.Unmarshal(data, v)
+	default:
+		return fmt.Errorf("cannot unmarshal: unsuported file %s", ext)
```

Comment:
as you're detecting the input based on the file extension, you could default the output based on this too i.e. if I input my step and test data as yaml, it's highly likely that's how I'd like my output too (currently it will be json)

Reply:
My implementation is using flyte's default which is json

Reply:
It might be the 'default', but it's a weird/poor user experience!